From b84b0d9031be82fc2eead408f31de3528c269ae8 Mon Sep 17 00:00:00 2001
From: GP Orcullo <kinsamanka@gmail.com>
Date: Sun, 5 Nov 2023 07:33:28 +0000
Subject: [PATCH 2/3] sophgo: add ethernet driver

Note: will not work as mainline now has an additional driver and the old
code added before is no longer added with this patch.

extracted from OEM supplied sources

Upstream-Status: Pending
---
 .../boot/dts/sophgo/cv1800b-milkv-duo.dts     |  4 ++
 arch/riscv/boot/dts/sophgo/cv180x.dtsi        | 52 +++++++++++++++++++
 2 files changed, 56 insertions(+)

diff --git a/arch/riscv/boot/dts/sophgo/cv1800b-milkv-duo.dts b/arch/riscv/boot/dts/sophgo/cv1800b-milkv-duo.dts
index 9feb520eaec4..a53ceb0ddb8e 100644
--- a/arch/riscv/boot/dts/sophgo/cv1800b-milkv-duo.dts
+++ b/arch/riscv/boot/dts/sophgo/cv1800b-milkv-duo.dts
@@ -35,6 +35,10 @@ coprocessor_rtos: region@83f40000 {
 	};
 };
 
+&emac0 {
+	status = "okay";
+};
+
 &osc {
 	clock-frequency = <25000000>;
 };
diff --git a/arch/riscv/boot/dts/sophgo/cv180x.dtsi b/arch/riscv/boot/dts/sophgo/cv180x.dtsi
index e9c8e72229e5..daf8af5a9a70 100644
--- a/arch/riscv/boot/dts/sophgo/cv180x.dtsi
+++ b/arch/riscv/boot/dts/sophgo/cv180x.dtsi
@@ -18,6 +18,20 @@ osc: oscillator {
 		#clock-cells = <0>;
 	};
 
+	clk_csr: clk_csr {
+		clock-output-names = "eth_csrclk";
+		clock-frequency = <250000000>;
+		#clock-cells = <0>;
+		compatible = "fixed-clock";
+	};
+
+	clk_ptp: clk_ptp {
+		clock-output-names = "eth_ptpclk";
+		clock-frequency = <50000000>;
+		#clock-cells = <0>;
+		compatible = "fixed-clock";
+	};
+
 	soc {
 		compatible = "simple-bus";
 		#address-cells = <1>;
@@ -312,6 +326,44 @@ restart: restart-controller@5025000 {
 			compatible = "sophgo,restart";
 			reg = <0x05025000 0x2000>;
 		};
+
+		stmmac_axi_setup: stmmac-axi-config {
+			snps,wr_osr_lmt = <1>;
+			snps,rd_osr_lmt = <2>;
+			snps,blen = <4 8 16 0 0 0 0>;
+		};
+
+		mtl_rx_setup: rx-queues-config {
+			snps,rx-queues-to-use = <1>;
+			queue0 {};
+		};
+
+		mtl_tx_setup: tx-queues-config {
+			snps,tx-queues-to-use = <1>;
+			queue0 {};
+		};
+
+		emac0: ethernet@4070000 {
+			compatible = "sophgo,ethernet";
+			reg = <0x4070000 0x10000>;
+			clock-names = "stmmaceth", "ptp_ref";
+			clocks = <&clk_csr>, <&clk_ptp>;
+			tx-fifo-depth = <8192>;
+			rx-fifo-depth = <8192>;
+			snps,multicast-filter-bins = <0>;
+			snps,perfect-filter-entries = <1>;
+			snps,txpbl = <8>;
+			snps,rxpbl = <8>;
+			snps,aal;
+			snps,axi-config = <&stmmac_axi_setup>;
+			snps,mtl-rx-config = <&mtl_rx_setup>;
+			snps,mtl-tx-config = <&mtl_tx_setup>;
+			phy-mode = "rmii";
+			interrupt-names = "macirq";
+			interrupts = <31 IRQ_TYPE_LEVEL_HIGH>;
+			interrupt-parent = <&plic>;
+			status = "disabled";
+		};
 	};
 
 };
-- 
2.49.0

